File: EcoSwapAndroid/src/main/java/com/darrenthiores/ecoswap/android/presentation/message/MessageScreen.kt
package com.darrenthiores.ecoswap.android.presentation.message

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material.Scaffold
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import com.darrenthiores.ecoswap.android.presentation.message.components.FriendMessageItem
import com.darrenthiores.ecoswap.android.presentation.message.components.MessageBottomBar
import com.darrenthiores.ecoswap.android.presentation.message.components.MessageHeader
import com.darrenthiores.ecoswap.android.presentation.message.components.PersonalMessageItem
import com.darrenthiores.ecoswap.presentation.message.MessageEvent
import com.darrenthiores.ecoswap.presentation.message.MessageState

@Composable
fun MessageScreen(
    state: MessageState,
    onEvent: (MessageEvent) -> Unit,
    onSettingClick: () -> Unit,
    onBackClick: () -> Unit
) {
    Scaffold(
        topBar = {
            MessageHeader(
                text = state.user?.name ?: "",
                onBackClick = onBackClick,
                onSettingClick = onSettingClick
            )
        },
        bottomBar = {
            MessageBottomBar(
                message = state.newMessage,
                onMessageChange = {
                    onEvent(
                        MessageEvent.OnMessageChange(it)
                    )
                },
                onAddPhotoClick = {  },
                onSendClick = {
                    onEvent(MessageEvent.SendMessage)
                }
            )
        }
    ) { paddingValues ->
        LazyColumn(
            modifier = Modifier
                .fillMaxSize()
                .padding(paddingValues),
            contentPadding = PaddingValues(16.dp),
            verticalArrangement = Arrangement.spacedBy(8.dp),
            reverseLayout = true
        ) {
            items(
                items = state.messages,
                key = { message -> message.id }
            ) { message ->
                if (message.sentToId == state.user?.id) {
                    PersonalMessageItem(
                        text = message.content,
                        username = message.sentFromUsername,
                        userImgUrl = message.sentFromImageUrl
                    )
                } else {
                    FriendMessageItem(
                        text = message.content,
                        username = message.sentToUsername,
                        userImgUrl = message.sentToImageUrl
                    )
                }
            }
        }
    }
}

File: EcoSwapiOS/EcoSwapiOS/Presentation/Message/MessageScreen.swift
//
//  MessageScreen.swift
//  EcoSwapiOS
//
//  Created by Darren Thiores on 08/01/24.
//  Copyright © 2024 orgName. All rights reserved.
//

import SwiftUI
import shared

struct MessageScreen: View {
    let userId: String
    
    @Environment(\.dismiss) private var dismiss
    @StateObject private var viewModel = IosMessageViewModel()
    
    var body: some View {
        GeometryReader { geo in
            VStack(spacing: 0) {
                MessageHeader(
                    title: viewModel.state.user?.name ?? "null",
                    onBackClick: {
                        dismiss()
                    },
                    onSettingClick: {  }
                )
                
                ScrollViewReader { scrollView in
                    ReversedScrollView(
                        .vertical,
                        proxy: geo
                    ) {
                        Spacer()
                            .frame(height: 16)
                        
                        ForEach(viewModel.state.messages.reversed(), id: \.id) { message in
                            if message.sentToId == userId {
                                PersonalMessageItem(
                                    text: message.content,
                                    username: message.sentFromUsername,
                                    userImgUrl: message.sentFromImageUrl,
                                    geo: geo
                                )
                            } else {
                                FriendMessageItem(
                                    text: message.content,
                                    username: message.sentToUsername,
                                    userImgUrl: message.sentToImageUrl,
                                    geo: geo
                                )
                            }
                        }
                        
                        Spacer()
                            .frame(height: 16)
                    }
                    .padding(.horizontal, 16)
                    .onChange(of: viewModel.state.messages) { messages in
                        if let message = messages.first {
                            withAnimation(.default) {
                                scrollView
                                    .scrollTo(message.id)
                            }
                        }
                    }
                }
                
                MessageBottomBar(
                    message: Binding(
                        get: {
                            viewModel.state.newMessage
                        },
                        set: {
                            viewModel.onEvent(
                                event: .OnMessageChange(newText: $0)
                            )
                        }
                    ),
                    onAddPhotoClick: {  },
                    onSendClick: {
                        viewModel.onEvent(
                            event: .SendMessage()
                        )
                    }
                )
            }
        }
        .navigationBarBackButtonHidden()
        .navigationBarTitleDisplayMode(.inline)
        .toolbar(.hidden, for: .tabBar)
        .onAppear {
            viewModel.startObserving()
            viewModel.initFetch(userId: userId)
        }
        .onDisappear {
            viewModel.dispose()
        }
    }
}

#Preview {
    MessageScreen(
        userId: "1"
    )
}
